### PR TITLE
Fix S2583 FP: Should not raise on double condition

### LINKED ISSUE #8470: Fix S2583 FP: Should not raise on `double` condition
### Description

Rule S2583 does not correctly identify all possible values of a double and incorrectly assumes a double cannot fall between (0, 1).

### Repro steps

		
```cs
double t = 0.5;
if (t <= 0) {
	return "a";
}

if (t >= 1) {  // S2583 is flagged. Assumes t must be >= 1
	return "b";
}

return "c";
```

### Expected behavior

The issue should not be flagged since t can be (0, 1).

### Actual behavior

S2583 is flagged.

### Known workarounds

Please provide a description of any known workarounds.

### Related information

* C# version: 10.0
* Visual Studio version: 17.8.2
* dotnet version : net6.0
* Operating System: Windows 11 Home 10.0.22000 Build 22000

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
Fixes #8470

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.cs [hunk 1]
```diff
@@ -97,16 +97,24 @@ private static ProgramState LearnBranchingNumberConstraint(ProgramState state, I
         var kind = falseBranch ? Opposite(binary.OperatorKind) : binary.OperatorKind;
         var leftNumber = state[binary.LeftOperand]?.Constraint<NumberConstraint>();
         var rightNumber = state[binary.RightOperand]?.Constraint<NumberConstraint>();
-        if (rightNumber is not null && !binary.LeftOperand.ConstantValue.HasValue && binary.LeftOperand.TrackedSymbol(state) is { } leftSymbol)
+        if (rightNumber is not null && OperandSymbol(binary.LeftOperand) is { } leftSymbol)
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.cs:100]: @cristian-ambrosini-sonarsource Don't forget to ask me for dedicated engine UT once you'll be reviewing this change :D

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.cs [hunk 2]
```diff
@@ -97,16 +97,24 @@ private static ProgramState LearnBranchingNumberConstraint(ProgramState state, I
...
         {
             state = LearnBranching(rightSymbol, rightNumber, Flip(kind), leftNumber);
         }
         return state;
 
+        ISymbol OperandSymbol(IOperation operand) =>
+            !operand.ConstantValue.HasValue
+            && operand.TrackedSymbol(state) is { } symbol
+            && symbol.GetSymbolType() is INamedTypeSymbol type
+            && (type.IsAny(KnownType.IntegralNumbersIncludingNative) || type.IsNullableOfAny(KnownType.IntegralNumbersIncludingNative))
```
Comment (cristian-ambrosini-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.cs:114]: Is `type.IsNullableOfAny(KnownType.IntegralNumbersIncludingNative)` covered by test cases?

**Review by cristian-ambrosini-sonarsource — CHANGES_REQUESTED** [review]
Left few comments, half of my brain is fried because of #8484 so I might have missed something, I would like to take another look tomorrow

File: analyzers/src/SonarAnalyzer.Common/Helpers/TypeHelper.cs [hunk 3]
```diff
@@ -205,4 +210,9 @@ symbol switch
             ITypeSymbol x => x,
             _ => null,
         };
+
+    private static ITypeSymbol NullableTypeArgument(ITypeSymbol type) =>
+        type is INamedTypeSymbol namedType && namedType.OriginalDefinition.Is(KnownType.System_Nullable_T)
+            ? namedType.TypeArguments[0]
```
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Helpers/TypeHelper.cs:216]: It's `System_Nullable_T`, and that one has exactly one T by-design.

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.cs (hunk 2 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.cs:114]: Many SE UTs failed without it.

**Review by cristian-ambrosini-sonarsource — APPROVED** [review]
LGTM!

**Pull request merged**